frontend/src/main/scala/bloop/engine/Interpreter.scala

```diff
@@ -1,41 +1,53 @@
 package bloop.engine
 
 import bloop.cli.{CliOptions, Commands, CommonOptions, ExitStatus}
+import bloop.io.SourceWatcher
 import bloop.io.Timer.timed
 import bloop.reporter.ReporterConfig
 import bloop.engine.tasks.{CompileTasks, TestTasks}
 import bloop.Project
+import bloop.logging.BloopLogger
 
 object Interpreter {
-  def execute(action: Action, state: State): State = {
-    import state.logger
-    def logAndTime[T](cliOptions: CliOptions, action: => T): T =
-      logger.verboseIf(cliOptions.verbose) { timed(state.logger) { action } }
+  def execute(action: Action, state0: State): State = {
+    def logAndTime[T](state: State, cliOptions: CliOptions, action: => T): T =
+      state.logger.verboseIf(cliOptions.verbose) { timed(state.logger) { action } }
+    def updateState(state: State, commonOptions: CommonOptions): State = {
+      State.updateLogger(state.logger, commonOptions)
+      val logger = BloopLogger(state.logger.name)
+      state.copy(commonOptions = commonOptions, logger = logger)
+    }
+
     action match {
-      case Exit(exitStatus) if state.status.isOk => state.copy(status = exitStatus)
-      case Exit(exitStatus) => state
+      case Exit(exitStatus) if state0.status.isOk => state0.mergeStatus(exitStatus)
+      case Exit(exitStatus) => state0
```

Comment:
Couldn't we just have `case Exit(exitStatus) => state0.mergeStatus(exitStatus)` ?

Reply:
No, because the implementation of mergeStatus is very naive at this point and has to improve. When there are two errors, it's best to at least report the first one and not a weird mix for two. I plan to improve this in a future PR